File: SoapWebApi/Services/MapService.cs
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Processing;
using SoapWebApi.Interfaces;
using SoapWebApi.Models;

namespace SoapWebApi.Services;

public class MapService : IMapService
{
    private readonly ILogger<MapService> _logger;

    // Anchor bbox for the single base image 1000x1000
    // Top-left geo anchor: (Lat 54.188424, Lon 19.387526)
    // Bottom-right geo anchor: (Lat 54.164336, Lon 19.428238)
    private const double ANCHOR_MIN_LAT = 54.164336; // bottom edge
    private const double ANCHOR_MAX_LAT = 54.188424; // top edge
    private const double ANCHOR_MIN_LON = 19.387526; // left edge
    private const double ANCHOR_MAX_LON = 19.428238; // right edge

    public MapService(ILogger<MapService> logger)
    {
        _logger = logger;
        _logger.LogInformation("MapService initialized");
    }

    public string Ping()
    {
        _logger.LogInformation("Ping method called");
        return $"Map Service is running. Time: {DateTime.Now}";
    }

    public MapResponse GetMapByPixelCoordinates(PixelMapRequest request)
    {
        try
        {
            // Defensive null checks to avoid NRE when deserialization fails
            if (request == null)
            {
                _logger.LogWarning("GetMapByPixelCoordinates: request is null (deserialization failed)");
                return new MapResponse
                {
                    Success = false,
                    Message = "Invalid request: request is null"
                };
            }

            if (request.TopLeft == null || request.BottomRight == null)
            {
                _logger.LogWarning(
                    "GetMapByPixelCoordinates: TopLeft or BottomRight is null. Likely XML namespace mismatch in client request.");
                return new MapResponse
                {
                    Success = false,
                    Message = "Invalid request: TopLeft and BottomRight pixel coordinates are required"
                };
            }

            _logger.LogInformation(
                $"GetMapByPixelCoordinates: TopLeft({request.TopLeft.X}, {request.TopLeft.Y}), BottomRight({request.BottomRight.X}, {request.BottomRight.Y})");

            if (request.ImageData == null || request.ImageData.Length == 0)
            {
                return new MapResponse
                {
                    Success = false,
                    Message = "Image data is required"
                };
            }

            using var image = Image.Load(request.ImageData);

            if (!ValidatePixelCoordinates(request, image.Width, image.Height))
            {
                return new MapResponse
                {
                    Success = false,
                    Message =
                        $"Invalid pixel coordinates. Image size: {image.Width.ToString()}x{image.Height.ToString()}"
                };
            }

            int width = request.BottomRight.X - request.TopLeft.X;
            int height = request.BottomRight.Y - request.TopLeft.Y;

            image.Mutate(x => x.Crop(new Rectangle(
                request.TopLeft.X,
                request.TopLeft.Y,
                width,
                height
            )));

            using var memoryStream = new MemoryStream();
            image.SaveAsJpeg(memoryStream);

            _logger.LogInformation($"Successfully created map crop: {width}x{height}");

            return new MapResponse
            {
                Success = true,
                ImageData = memoryStream.ToArray(),
                ContentType = "image/jpeg",
                Width = width,
                Height = height,
                Message = "Success"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to process map image");
            return new MapResponse
            {
                Success = false,
                Message = $"Error: {ex.Message}"
            };
        }
    }

    public MapResponse GetMapByGeoCoordinates(GeoMapRequest request)
    {
        try
        {
            // Defensive null checks to avoid NRE when deserialization fails or namespaces mismatch
            if (request == null)
            {
                _logger.LogWarning("GetMapByGeoCoordinates: request is null (deserialization failed)");
                return new MapResponse
                {
                    Success = false,
                    Message = "Invalid request: request is null"
                };
            }

            if (request.TopLeft == null || request.BottomRight == null)
            {
                _logger.LogWarning(
                    "GetMapByGeoCoordinates: TopLeft or BottomRight is null. Likely XML namespace mismatch in client request.");
                return new MapResponse
                {
                    Success = false,
                    Message = "Invalid request: TopLeft and BottomRight geo coordinates are required"
                };
            }

            _logger.LogInformation(
                $"GetMapByGeoCoordinates: TopLeft(Lat: {request.TopLeft.Latitude}, Lon: {request.TopLeft.Longitude}), BottomRight(Lat: {request.BottomRight.Latitude}, Lon: {request.BottomRight.Longitude})");

            if (request.ImageData == null || request.ImageData.Length == 0)
            {
                return new MapResponse
                {
                    Success = false,
                    Message = "Image data is required"
                };
            }

            if (!ValidateGeoCoordinates(request))
            {
                return new MapResponse
                {
                    Success = false,
                    Message = "Coordinates exceed allowed bounds"
                };
            }

            using var image = Image.Load(request.ImageData);

            var pixelTopLeft = GeoToPixel(
                request.TopLeft.Latitude,
                request.TopLeft.Longitude,
                image.Width,
                image.Height
            );

            var pixelBottomRight = GeoToPixel(
                request.BottomRight.Latitude,
                request.BottomRight.Longitude,
                image.Width,
                image.Height
            );

            _logger.LogInformation(
                $"Converted to pixels: TopLeft({pixelTopLeft.X}, {pixelTopLeft.Y}), BottomRight({pixelBottomRight.X}, {pixelBottomRight.Y})");

            // Normalize rectangle if needed to be safe
            var left = Math.Min(pixelTopLeft.X, pixelBottomRight.X);
            var right = Math.Max(pixelTopLeft.X, pixelBottomRight.X);
            var top = Math.Min(pixelTopLeft.Y, pixelBottomRight.Y);
            var bottom = Math.Max(pixelTopLeft.Y, pixelBottomRight.Y);

            var pixelRequest = new PixelMapRequest
            {
                TopLeft = new PixelCoordinates { X = left, Y = top },
                BottomRight = new PixelCoordinates { X = right, Y = bottom },
                ImageData = request.ImageData
            };

            if (!ValidatePixelCoordinates(pixelRequest, image.Width, image.Height))
            {
                return new MapResponse
                {
                    Success = false,
                    Message = "Invalid pixel coordinates"
                };
            }

            int width = right - left;
            int height = bottom - top;

            image.Mutate(x => x.Crop(new Rectangle(
                left,
                top,
                width,
                height
            )));

            using var memoryStream = new MemoryStream();
            image.SaveAsJpeg(memoryStream);

            _logger.LogInformation($"Successfully created map crop from geo coordinates: {width}x{height}");

            return new MapResponse
            {
                Success = true,
                ImageData = memoryStream.ToArray(),
                ContentType = "image/jpeg",
                Width = width,
                Height = height,
                Message = "Success"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to process map image from geo coordinates");
            return new MapResponse
            {
                Success = false,
                Message = $"Error: {ex.Message}"
            };
        }
    }

    private PixelCoordinates GeoToPixel(double latitude, double longitude, int imageWidth, int imageHeight)
    {
        // Linear mapping based on anchors
        double normalizedLon = (longitude - ANCHOR_MIN_LON) / (ANCHOR_MAX_LON - ANCHOR_MIN_LON);
        // y-axis inversion for image coordinates (top=0): higher latitude -> smaller y
        double normalizedLat = (ANCHOR_MAX_LAT - latitude) / (ANCHOR_MAX_LAT - ANCHOR_MIN_LAT);

        normalizedLon = Math.Max(0, Math.Min(1, normalizedLon));
        normalizedLat = Math.Max(0, Math.Min(1, normalizedLat));

        return new PixelCoordinates
        {
            X = (int)(normalizedLon * imageWidth),
            Y = (int)(normalizedLat * imageHeight)
        };
    }

    private bool ValidatePixelCoordinates(PixelMapRequest request, int imageWidth, int imageHeight)
    {
        bool isValid = request.TopLeft.X >= 0 &&
                       request.TopLeft.Y >= 0 &&
                       request.BottomRight.X <= imageWidth &&
                       request.BottomRight.Y <= imageHeight &&
                       request.TopLeft.X < request.BottomRight.X &&
                       request.TopLeft.Y < request.BottomRight.Y;

        if (!isValid)
        {
            _logger.LogWarning(
                $"Invalid pixel coordinates. Image: {imageWidth}x{imageHeight}, Request: TopLeft({request.TopLeft.X},{request.TopLeft.Y}), BottomRight({request.BottomRight.X},{request.BottomRight.Y})");
        }

        return isValid;
    }

    private bool ValidateGeoCoordinates(GeoMapRequest request)
    {
        bool isValid = request.TopLeft.Latitude >= ANCHOR_MIN_LAT &&
                       request.TopLeft.Latitude <= ANCHOR_MAX_LAT &&
                       request.BottomRight.Latitude >= ANCHOR_MIN_LAT &&
                       request.BottomRight.Latitude <= ANCHOR_MAX_LAT &&
                       request.TopLeft.Longitude >= ANCHOR_MIN_LON &&
                       request.TopLeft.Longitude <= ANCHOR_MAX_LON &&
                       request.BottomRight.Longitude >= ANCHOR_MIN_LON &&
                       request.BottomRight.Longitude <= ANCHOR_MAX_LON &&
                       request.TopLeft.Latitude > request.BottomRight.Latitude &&
                       request.TopLeft.Longitude < request.BottomRight.Longitude;

        if (!isValid)
        {
            _logger.LogWarning(
                $"Invalid geo coordinates. Allowed bounds: Lat({ANCHOR_MIN_LAT}-{ANCHOR_MAX_LAT}), Lon({ANCHOR_MIN_LON}-{ANCHOR_MAX_LON}), Request: TopLeft(Lat:{request.TopLeft.Latitude}, Lon:{request.TopLeft.Longitude}), BottomRight(Lat:{request.BottomRight.Latitude}, Lon:{request.BottomRight.Longitude})");
        }

        return isValid;
    }
}

File: stm_mobile_app/lib/services/map_service.dart
import 'dart:convert';
import 'dart:async';
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:xml/xml.dart' as xml;

class MapServiceClient {
  MapServiceClient({required this.baseUrl});

  /// Base URL of the SOAP service, e.g. http://localhost:5000
  final String baseUrl;

  // Anchors for geo→pixel conversion (confirmed by you)
  static const double _anchorMinLat = 54.164336; // bottom edge (south)
  static const double _anchorMaxLat = 54.188424; // top edge (north)
  static const double _anchorMinLon = 19.387526; // left edge (west)
  static const double _anchorMaxLon = 19.428238; // right edge (east)

  // Default base image size (confirmed single PNG 1000x1000)
  static const int _defaultImageWidth = 1000;
  static const int _defaultImageHeight = 1000;

  /// Calls SOAP method GetMapByPixelCoordinates.
  /// [topLeftX],[topLeftY],[bottomRightX],[bottomRightY] define the crop rectangle.
  /// [imageBytes] is the source image to crop on the backend.
  /// Returns decoded JPEG bytes of the cropped image on success.
  Future<Uint8List> getMapByPixelCoordinates({
    required int topLeftX,
    required int topLeftY,
    required int bottomRightX,
    required int bottomRightY,
    required Uint8List imageBytes,
  }) async {
    final endpoint = Uri.parse('$baseUrl/MapService.svc');

    final imageBase64 = base64Encode(imageBytes);

    const svcNs = 'http://mapservice.soap.api/2024';
    const modelNs = 'http://schemas.datacontract.org/2004/07/SoapWebApi.Models';

    // Build SOAP envelope matching CoreWCF expectations:
    // - Operation wrapper in service namespace (svcNs)
    // - Parameter element (request) in model namespace (modelNs)
    // - Data members in model namespace (modelNs)
    final envelope = '''
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns="$svcNs" xmlns:dt="$modelNs">
  <soapenv:Header/>
  <soapenv:Body>
    <ns:GetMapByPixelCoordinates>
      <ns:request>
        <dt:BottomRight>
          <dt:X>$bottomRightX</dt:X>
          <dt:Y>$bottomRightY</dt:Y>
        </dt:BottomRight>
        <dt:ImageData>$imageBase64</dt:ImageData>
        <dt:TopLeft>
          <dt:X>$topLeftX</dt:X>
          <dt:Y>$topLeftY</dt:Y>
        </dt:TopLeft>
      </ns:request>
    </ns:GetMapByPixelCoordinates>
  </soapenv:Body>
</soapenv:Envelope>
''';

    final headers = <String, String>{
      'Content-Type': 'text/xml; charset=utf-8',
      'SOAPAction': 'http://mapservice.soap.api/2024/IMapService/GetMapByPixelCoordinates',
    };

    final resp = await http
        .post(endpoint, headers: headers, body: envelope)
        .timeout(const Duration(seconds: 15));
    if (resp.statusCode < 200 || resp.statusCode >= 300) {
      throw Exception('HTTP ${resp.statusCode}: ${resp.body}');
    }

    final doc = xml.XmlDocument.parse(resp.body);

    // Helper to find first element by local name (ignore namespaces)
    xml.XmlElement? _firstByLocal(String local) {
      final it = doc.descendants.whereType<xml.XmlElement>();
      for (final e in it) {
        if (e.name.local == local) return e;
      }
      return null;
    }

    // Try to read ImageData regardless of Success flag (some services may omit Success)
    final imageNode = _firstByLocal('ImageData') ?? doc.findAllElements('ImageData').firstOrNull;

    // Read Success if present
    final successNode = _firstByLocal('Success') ?? doc.findAllElements('Success').firstOrNull;
    final isSuccess = successNode?.innerText.trim().toLowerCase() == 'true';

    // If we have image data, prefer returning it even if Success flag is missing/false
    if (imageNode != null) {
      final imgB64 = imageNode.innerText.trim();
      if (imgB64.isNotEmpty) {
        return base64Decode(imgB64);
      }
    }

    // Otherwise, if explicitly successful but no image found, treat as error
    if (isSuccess == true) {
      throw Exception('Service error: No ImageData in response');
    }

    // Extract message (namespace-agnostic)
    final messageNode = _firstByLocal('Message') ?? doc.findAllElements('Message').firstOrNull;
    final msg = messageNode?.innerText.trim().isNotEmpty == true
        ? messageNode!.innerText.trim()
        : 'Unknown error';

    throw Exception('Service error: $msg');
  }

  /// Calls SOAP method GetMapByGeoCoordinates.
  /// [topLeftLat],[topLeftLon],[bottomRightLat],[bottomRightLon] define geo rectangle.
  /// [imageBytes] is the source image used by backend.
  /// Returns decoded JPEG bytes on success.
  Future<Uint8List> getMapByGeoCoordinates({
    required double topLeftLat,
    required double topLeftLon,
    required double bottomRightLat,
    required double bottomRightLon,
    required Uint8List imageBytes,
  }) async {
    final endpoint = Uri.parse('$baseUrl/MapService.svc');

    final imageBase64 = base64Encode(imageBytes);

    const svcNs = 'http://mapservice.soap.api/2024';
    const modelNs = 'http://schemas.datacontract.org/2004/07/SoapWebApi.Models';

    final envelope = '''
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns="$svcNs" xmlns:dt="$modelNs">
  <soapenv:Header/>
  <soapenv:Body>
    <ns:GetMapByGeoCoordinates>
      <ns:request>
        <dt:TopLeft>
          <dt:Latitude>$topLeftLat</dt:Latitude>
          <dt:Longitude>$topLeftLon</dt:Longitude>
        </dt:TopLeft>
        <dt:BottomRight>
          <dt:Latitude>$bottomRightLat</dt:Latitude>
          <dt:Longitude>$bottomRightLon</dt:Longitude>
        </dt:BottomRight>
        <dt:ImageData>$imageBase64</dt:ImageData>
      </ns:request>
    </ns:GetMapByGeoCoordinates>
  </soapenv:Body>
</soapenv:Envelope>
''';

    final headers = <String, String>{
      'Content-Type': 'text/xml; charset=utf-8',
      'SOAPAction': 'http://mapservice.soap.api/2024/IMapService/GetMapByGeoCoordinates',
    };

    final resp = await http
        .post(endpoint, headers: headers, body: envelope)
        .timeout(const Duration(seconds: 15));
    print(resp.body);
    if (resp.statusCode < 200 || resp.statusCode >= 300) {
      throw Exception('HTTP ${resp.statusCode}: ${resp.body}');
    }

    final doc = xml.XmlDocument.parse(resp.body);

    xml.XmlElement? _firstByLocal(String local) {
      final it = doc.descendants.whereType<xml.XmlElement>();
      for (final e in it) {
        if (e.name.local == local) return e;
      }
      return null;
    }

    final imageNode = _firstByLocal('ImageData') ?? doc.findAllElements('ImageData').firstOrNull;
    if (imageNode != null) {
      final b64 = imageNode.innerText.trim();
      if (b64.isNotEmpty) {
        return base64Decode(b64);
      }
    }

    final messageNode = _firstByLocal('Message') ?? doc.findAllElements('Message').firstOrNull;
    final msg = messageNode?.innerText.trim().isNotEmpty == true
        ? messageNode!.innerText.trim()
        : 'Unknown error';
    throw Exception('Service error: $msg');
  }

  /// Client-side geo→pixel conversion using the confirmed anchors
  /// and then call the working pixel endpoint. This bypasses SOAP
  /// deserialization fragility for the Geo operation.
  Future<Uint8List> getMapByGeoViaPixel({
    required double topLeftLat,
    required double topLeftLon,
    required double bottomRightLat,
    required double bottomRightLon,
    required Uint8List imageBytes,
    int imageWidth = _defaultImageWidth,
    int imageHeight = _defaultImageHeight,
  }) async {
    // Convert both corners to image pixel space (origin at top-left)
    final tl = _geoToPixel(topLeftLat, topLeftLon, imageWidth, imageHeight);
    final br = _geoToPixel(bottomRightLat, bottomRightLon, imageWidth, imageHeight);

    // Normalize rectangle: ensure left<right and top<bottom
    final left = tl.$1 < br.$1 ? tl.$1 : br.$1;
    final right = tl.$1 < br.$1 ? br.$1 : tl.$1;
    final top = tl.$2 < br.$2 ? tl.$2 : br.$2;
    final bottom = tl.$2 < br.$2 ? br.$2 : tl.$2;

    return getMapByPixelCoordinates(
      topLeftX: left,
      topLeftY: top,
      bottomRightX: right,
      bottomRightY: bottom,
      imageBytes: imageBytes,
    );
  }

  /// Returns (x,y) in image pixel space (top-left origin)
  /// Using linear interpolation between the geo anchors.
  /// Values are clamped to [0..W]x[0..H].
  (int, int) _geoToPixel(double lat, double lon, int w, int h) {
    // Normalize lon to [0,1] left→right
    double nLon = (lon - _anchorMinLon) / (_anchorMaxLon - _anchorMinLon);
    // Normalize lat to [0,1] top→bottom (invert because lat increases northward)
    double nLat = (_anchorMaxLat - lat) / (_anchorMaxLat - _anchorMinLat);

    // Clamp
    nLon = nLon.clamp(0.0, 1.0);
    nLat = nLat.clamp(0.0, 1.0);

    final x = (nLon * w).toInt();
    final y = (nLat * h).toInt();
    return (x, y);
  }
}

extension _FirstOrNullExtension<E> on Iterable<E> {
  E? get firstOrNull => isEmpty ? null : first;
}
